examples/nine-pt-stencil-solver/nine-pt-stencil-solver.cpp

```diff
@@ -0,0 +1,301 @@
+// points.
+void generate_stencil_matrix(int dp, int *row_ptrs, int *col_idxs,
+                             double *values)
+{
+    int pos = 0;
+    size_t dp_2 = dp * dp;
+    row_ptrs[0] = pos;
+    for (int k = 0; k < dp; ++k) {
+        for (int i = 0; i < dp; ++i) {
+            const size_t index = i + k * dp;
+            for (int j = -1; j <= 1; ++j) {
+                for (int l = -1; l <= 1; ++l) {
+                    const int64_t offset = l + 1 + 3 * (j + 1);
+                    if ((k + j) >= 0 && (k + j) < dp && (i + l) >= 0 &&
+                        (i + l) < dp) {
+                        values[pos] = coefs[offset];
+                        col_idxs[pos] = index + l + dp * j;
+                        ++pos;
+                    }
+                }
+            }
+            row_ptrs[index + 1] = pos;
+        }
+    }
+}
+
+
+// Generates the RHS vector given `f` and the boundary conditions.
+template <typename Closure, typename ClosureT>
+void generate_rhs(int dp, Closure f, ClosureT u, double *rhs)
+{
+    const size_t dp_2 = dp * dp;
+    const auto h = 1.0 / (dp + 1.0);
+    for (int i = 0; i < dp; ++i) {
+        const auto yi = (i + 1) * h;
+        for (int j = 0; j < dp; ++j) {
+            const auto xi = (j + 1) * h;
+            const auto index = i * dp + j;
+            rhs[index] = -f(xi, yi) * h * h;
+        }
+    }
+
+    // Iterating over the edges to add boundary values
+    // and adding the overlapping 3x1 to the rhs
+    for (size_t i = 0; i < dp; ++i) {
+        const auto xi = (i + 1) * h;
+        const auto index_top = i;
+        const auto index_bot = i + dp * (dp - 1);
+
+        rhs[index_top] -= u(xi - h, 0.0) * coefs[0];
+        rhs[index_top] -= u(xi, 0.0) * coefs[1];
+        rhs[index_top] -= u(xi + h, 0.0) * coefs[2];
+
+        rhs[index_bot] -= u(xi - h, 1.0) * coefs[6];
+        rhs[index_bot] -= u(xi, 1.0) * coefs[7];
+        rhs[index_bot] -= u(xi + h, 1.0) * coefs[8];
+    }
+    for (size_t i = 0; i < dp; ++i) {
+        const auto yi = (i + 1) * h;
+        const auto index_left = i * dp;
+        const auto index_right = i * dp + (dp - 1);
+
+        rhs[index_left] -= u(0.0, yi - h) * coefs[0];
+        rhs[index_left] -= u(0.0, yi) * coefs[3];
+        rhs[index_left] -= u(0.0, yi + h) * coefs[6];
+
+        rhs[index_right] -= u(1.0, yi - h) * coefs[2];
+        rhs[index_right] -= u(1.0, yi) * coefs[5];
+        rhs[index_right] -= u(1.0, yi + h) * coefs[8];
+    }
+
+    // remove the double corner values
+
+    rhs[0] += u(0.0, 0.0) * coefs[0];
+    rhs[(dp - 1)] += u(1.0, 0.0) * coefs[2];
+    rhs[(dp - 1) * dp] += u(0.0, 1.0) * coefs[6];
+    rhs[dp * dp - 1] += u(1.0, 1.0) * coefs[8];
+}
+
+
+// Prints the solution `u`.
+void print_solution(int dp, const double *u)
+{
+    for (int i = 0; i < dp; ++i) {
+        for (int j = 0; j < dp; ++j) {
+            std::cout << u[i * dp + j] << ' ';
+        }
+        std::cout << std::endl;
```

Comment:
If the matrix is large, this might decrease performance significantly:
> Use of std::endl in place of '\n', encouraged by some sources, may significantly degrade output performance.

--- [cppreference](https://en.cppreference.com/w/cpp/io/manip/endl)
Since you have the `std::endl` at the end either way, you don't need to flush the buffer every single line.
```suggestion
        std::cout << '\n';
```